### PR TITLE
Signature map

### PR DESCRIPTION
**Pull request by privat**
Provide an abstraction `SingatureMap` to factorize (and simplify) the association between arguments and parameters (not that easy in the context of varargs).

This PR is also required by a future PR.

### CODE REVIEW

**Comment by privat** [comment]
reroll: update super in niti to try to pass tests

File: src/compiler/abstract_compiler.nit [hunk 1]
```diff
@@ -1146,40 +1146,43 @@ abstract class AbstractCompilerVisitor
...
-					vararg.add(e)
-				end
-				var elttype = msignature.mparameters[vararg_rank].mtype
+			var param = msignature.mparameters[i]
+			var j = map.map.get_or_null(i)
+			if j == null then
+				continue
+			end
+			if param.is_vararg and map.vararg_decl > 0 then
+				var vararg = [for k in [j..j+map.vararg_decl[ do exprs[k]]
```
Comment (Morriar) [inline src/compiler/abstract_compiler.nit]: Hum...

File: src/interpreter/naive_interpreter.nit [hunk 2]
```diff
@@ -370,42 +370,48 @@ class NaiveInterpreter
...
-				var e = self.expr(args[j])
-				if e == null then return null
-				res.add(e)
+			var param = msignature.mparameters[i]
+			var j = map.map.get_or_null(i)
+			if j == null then
+				continue
+			end
+			if param.is_vararg and map.vararg_decl > 0 then
+				var vararg = [for k in [j..j+map.vararg_decl[ do exprs[k]]
```
Comment (Morriar) [inline src/interpreter/naive_interpreter.nit]: It's definitely a "Hum..."

File: src/compiler/abstract_compiler.nit (hunk 1 above)
Comment (privat) [inline src/compiler/abstract_compiler.nit]:
~~~nit
var vararg = exprs.slice(j, map.vararg_decl)
                   ^ unknown slice method on Array
~~~

OK, I will add the method.

**Comment by Morriar** [comment]
+1 but there is something to do with those array comprehensions

**Comment by privat** [comment]
reroll: introduce `Array::sub` and use it instead of the two ugly array comprehensions

**Comment by xymus** [comment]
+1

**Comment by BlackMinou** [comment]
+1

**Pull request merged**